Auth test

mataa019 opened the pull request:
feat: Implement Tazama Auth Service Integration with RBAC
Summary
Integrated CMS with Tazama Auth Service for centralized authentication, role-based access control, and tenant isolation.

Key Features
Authentication: JWT-based login via Tazama Auth Service
Authorization: Role & permission-based access control with @roles() decorator
Tenant Isolation: Data segregation by tenantId from JWT claims
Audit Logging: Comprehensive tracking of all user actions and security events
Security: RS256 token validation, proper error handling, audit trails
Technical Implementation
AuthService, JwtStrategy, AuthController for authentication flow
RolesGuard with @roles() decorator for endpoint protection
AuditLogService for compliance logging with UUID tracking
Triage module integration with JWT authentication
Testing & Quality
80%+ Test Coverage for Auth and Audit modules
Unit tests for all services, controllers, guards, and strategies
ESLint + Prettier configuration with Husky pre-commit hooks
Comprehensive error handling and security validation tests
Acceptance Criteria Met
All story requirements fulfilled: centralized auth, RBAC, tenant isolation, audit logging, token management, and security controls.

Akku-Rawat committed:
feat(prisma): complete initial schema setup

Akku-Rawat committed:
feat(triage) :  manual alert triage

mataa019 committed:
Merge branch 'feat/triage-auto-trigger' into feature-auth

mataa019 committed:
Merge pull request #2 from tazama-lf/feature-auth

Feature auth

mataa019 committed:
feat:fixing the conflic merge

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat:adding auth service

mataa019 committed:
feat: Auth adding roles decorators

mataa019 committed:
feat: Unit Test for Auth and Audit

mataa019 committed:
feat: Test Coverage

mataa019 committed:
feat: Test Coverage

mataa019 committed:
feat:Test to Triage Module

mataa019 committed:
feat: Test Coverage for Triage Module

Abdullah-Siddiqui-Paysys commented on .husky/pre-commit:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
+
+npm run lint
```
Add  'npm test' in a new file named 'pre-push'

mataa019 replied:
Added a new pre-push file that runs npm test as requested. All tests now run automatically before pushing to the repository.

Abdullah-Siddiqui-Paysys commented on src/triage/triage.controller.ts:
```diff
@@ -1,13 +1,62 @@
+import { RolesGuard } from '../auth/roles.guard';
 
-@Controller('api/v1/triage/alerts')
+@Controller('api/v1/triage')
```
Can I suggest adding authentication guards at the controller level? Currently, any user can access all endpoints without authentication, which violates **Acceptance Criteria no.1** (Centralized Authentication):

```typescript
@Controller('api/v1/triage/alerts')
@UseGuards(AuthGuard('jwt'), RolesGuard) 
export class TriageController {

mataa019 replied:
I've added @UseGuards(AuthGuard('jwt'), RolesGuard) to the controller to enforce centralized authentication as per the acceptance criteria.

Abdullah-Siddiqui-Paysys commented on src/triage/triage.service.ts:
```diff
@@ -1,63 +1,156 @@
+  }
+
+  async updateAlertData(alertId: string, dto: UpdateAlertDto, userId: string) {
+    const alert = await this.prisma.alert.findUnique({
```
The `updateAlertData` method doesn't enforce tenant isolation - it can update alerts from any tenant.
Add tenant validation:
```typescript
const alert = await this.prisma.alert.findUnique({
  where: { 
    alert_id: alertId,
    tenant_id: tenantId
  },
});

mataa019 replied:
Tenant isolation has been added by updating the findUnique query to include tenant_id in the where clause as suggested. This now ensures alerts are only updated within the correct tenant scope.

Abdullah-Siddiqui-Paysys commented on src/triage/triage.service.ts:
```diff
@@ -1,63 +1,156 @@
-      created_at: new Date().toISOString(),
-    };
+  async manualCloseAlert(alertId: string, status: AlertStatus, userId: string) {
+    const alert = await this.prisma.alert.findUnique({
```
The `manualCloseAlert` method doesn't enforce tenant isolation - it can update alerts from any tenant.
Add tenant validation:

```
const alert = await this.prisma.alert.findUnique({
  where: { 
    alert_id: alertId,
    tenant_id: tenantId
  },
});
```

Abdullah-Siddiqui-Paysys commented on src/auth/auth.service.ts:
```diff
@@ -14,26 +12,28 @@ export class AuthService {
-        );
-        const token = typeof response.data === 'string'
+  async login(username: string, password: string) {
+    const authUrl = this.configService.get<string>('TAZAMA_AUTH_URL');
```
Please add a .env.template file that can be used via command: `cp .env.template .env`

mataa019 replied:
Added a .env.template file with all necessary environment variables. Can now be used with cp .env.template .env for setup consistency.

Abdullah-Siddiqui-Paysys commented on src/auth/auth.service.ts:
```diff
@@ -14,26 +12,28 @@ export class AuthService {
-      try {
-        const response = await firstValueFrom(
-          this.httpService.post(authUrl, { username, password })
-        );
```
No token refresh capability implemented.

Violates acceptance criteria no.6 - session and token expiry handling.

mataa019 replied:
The CMS now handles access token expiry

Abdullah-Siddiqui-Paysys commented on src/auth/jwt.strategy.ts:
```diff
@@ -2,7 +2,6 @@ import { Injectable } from '@nestjs/common';
 import { PassportStrategy } from '@nestjs/passport';
```
No token refresh capability.
Violates acceptance criteria no.6 - session and token expiry handling.

Abdullah-Siddiqui-Paysys commented on src/auth/auth.service.ts:
```diff
@@ -14,26 +12,28 @@ export class AuthService {
+            response.data?.user?.token;
+      this.logger.log('Login successful');
+      return { token };
+    } catch (error) {
```
All authentication failures return "Authentication failed" - no error categorization.
Distinguish credentials vs service issues.
Violates acceptance criteria no.7

mataa019 replied:
Authentication error handling has been updated to distinguish between invalid credentials (e.g. wrong username/password) and internal/service-related issues. Appropriate status codes and messages are now returned, fulfilling Acceptance Criteria no.7.

Abdullah-Siddiqui-Paysys commented on src/app.module.ts:
```diff
@@ -1,11 +1,11 @@
 import { Module } from '@nestjs/common';
 import { ConfigModule } from '@nestjs/config';
 import { AuditLogModule } from './audit/auditLog.module';
```
utlise `LoggerService` as mentinoed in the user story:

Every login, token validation, and permission denial must be logged in:
System logs for operational tracing
Audit logs for regulatory compliance.

mataa019 replied:
LoggerService is implemented across all relevant authentication flows. Login attempts, token validations, and permission denials are logged to both the system and audit logs, ensuring traceability and compliance with the user story requirements.

Akku-Rawat committed:
feat(triage): send alert for manual investigation

mataa019 committed:
feat: test

Akku-Rawat committed:
feat(triage): enforce tenant isolation for alert operations

- Restricted update, manual close, and investigation actions to the tenant that owns the alert.
- Prevents unauthorized access or manipulation of alerts across tenants.
- Throws NotFoundException when tenant mismatch occurs.

mataa019 committed:
feat: add the pre-push file

mataa019 committed:
feat: Authentication Guards Added at Controller Level

mataa019 committed:
feat: Authentication Guards Added at Controller Level

mataa019 committed:
feat: add a .env.template file that can be used via command: cp .env.template .env

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

Abdullah-Siddiqui-Paysys commented on test/app.module.spec.ts:
```diff
@@ -0,0 +1,32 @@
+import { Test, TestingModule } from '@nestjs/testing';
```
Make sure test coverage is above threshold value in all test files

mataa019 replied:
Done that

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
feat: token refresh functionality implemented

mataa019 committed:
Add .env.example (renamed from .env.template)

mataa019 committed:
feat: fixing the esLint and prettier errors

mataa019 committed:
Add .env.example (renamed from .env.template)

mataa019 committed:
Add .env.example (renamed from .env.template)

Akku-Rawat committed:
Merge branch 'Auth-test' into izyane

mwan9ilwa committed:
fixed coverage threshold

mwan9ilwa committed:
Test pre-commit hook with user decorator tests

mwan9ilwa committed:
Update Husky configuration to v10 format

Akku-Rawat committed:
Merge branch 'izyane' into Auth-Test

Abdullah-Siddiqui-Paysys reviewed (approved):
ok
